Fix crash when widgetspan does not produce a semantics node in render…

chunhtai opened the pull request:
… paragraph

## Description
The change adds semantics tags to mark the relationship between semantics nodes and their corresponding render object.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/69787

## Tests

I added the following tests:

see files

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I signed the [CLA].
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

chunhtai committed:
Fix crash when widgetspan does not produce a semantics node in render paragraph

chunhtai committed:
fix comments

chunhtai commented:
cc @mehmetf

goderbauer commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -5369,6 +5369,48 @@ class Flow extends MultiChildRenderObjectWidget {
   }
 }
 
+class _SemanticsTagWidget extends SingleChildRenderObjectWidget {
```
Can we just roll this functionality into the Semantics widget? That is, the Semantics widget should have a `tag` property to take a `SemanticsTag`.

chunhtai replied:
what is the correct behavior on the setting semantics tags in SemanticsNode? does it add the tag to itself or does it add the tags for the subtree only?

goderbauer replied:
What do you mean? We only have the `addTagForChildren` API, no?

chunhtai replied:
oh, yes, so the behavior should be adding tag for the semantics subtree except itself if it also introduce a semantics node?

goderbauer replied:
Yeah

dbrinegar commented on packages/flutter/lib/src/rendering/paragraph.dart:
```diff
@@ -956,6 +987,10 @@ class RenderParagraph extends RenderBox
       }
     }
+    // Makes sure we exhausted the list.
+    assert(childIndex == children.length);
```
There seem valid cases where childIndex < children.length.

In my test where this used to crash, I get childIndex: 2, children.length 7, placeholderIndex: 8 at this point, as a reference.

chunhtai replied:
That is unexpected, we have to annotate all the semantics children, or they may not have the correct semantics. Do you have a repro example?

mehmetf replied:
Does my complicated example help? That is pretty close to what customer:money is actually doing. https://github.com/flutter/flutter/issues/69787#issuecomment-722169455

goderbauer replied:
We may be missing the use case where a WidgetSpan is bubbling up more than one semantic node, something like:

```dart
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Directionality(
      textDirection: TextDirection.ltr,
      child: Center(
        child: Container(
          color: Colors.green,
          child: RichText(
            text: TextSpan(
              children: [
                TextSpan(
                  text: 'Start\n',
                  recognizer: TapGestureRecognizer()..onTap = () { },
                ),
                WidgetSpan(
                  child: Icon(
                    Icons.edit,
                    size: 16,
                  ),
                ),
                WidgetSpan(
                  child: Row(
                    children: [
                      Semantics(
                        container: true,
                        child: Icon(
                          Icons.edit,
                          size: 16,
                          semanticLabel: 'foo',
                        ),
                      ),
                      Semantics(
                        container: true,
                        child: Icon(
                          Icons.edit,
                          size: 16,
                          semanticLabel: 'bar',
                        ),
                      ),
                    ],
                  ),
                ),
                TextSpan(text: 'End'),
              ],
            ),
          ),
        ),
      ),
    );
  }
}
```

@chunhtai can you try this with your change and see if it works? There can probably be more than one semantics node in `children` for a given `placeholderIndex`.

chunhtai replied:
yes, this pr missed this use case, I will update it to make sure it works

dbrinegar replied:
Also feel free to ping me on chat if would be helpful, happy to walk through it or help dig. I don't have a simple setup, but have a snapshot with the above crash.

chunhtai committed:
fixed corner case, added test, refactored semantics widget

chunhtai commented:
Thanks for reviewing ! @@goderbauer @mehmetf @dbrinegar 
Addressed all of the issues, ready for another look

goderbauer commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -913,6 +914,12 @@ class SemanticsProperties extends DiagnosticableTree {
   /// on iOS and TalkBack on Android).
   final SemanticsSortKey? sortKey;
 
+  /// Adds additional information onto the semantics subtree.
+  ///
+  /// This property provides a way for rendering objects to annotate their
+  /// semantics children without directly access them.
```
Specifies a set of tags to be applied to the child SemanticsNodes of this widget.

The tag is added to all child SemanticsNodes that pass through the RenderObject corresponding to this widget while looking to be attached to a parent SemanticsNode.

Tags are used to communicate to a parent SemanticsNode that a child SemanticsNode was passed through a particular RenderObject. The parent can use this information to determine the shape of the semantics tree.

See also:

 * [SemanticsConfiguration.addTagForChildren], to which the tags provided here will be passed.

goderbauer commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -4072,6 +4074,17 @@ class RenderSemanticsAnnotations extends RenderProxyBox {
   }
 
+  /// Adds Semenatics tags to the semantics subtree.
+  Set<SemanticsTag>? get tagsForChildren => _tagsForChildren;
```
I wonder if we should simplify the API here and in the Semantics widget to allow only one tag instead of a set. If you need multiple tags (uncommon) you can still wrap your tree in multiple Semantics widgets.

goderbauer commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -913,6 +914,12 @@ class SemanticsProperties extends DiagnosticableTree {
+  ///
+  /// This property provides a way for rendering objects to annotate their
+  /// semantics children without directly access them.
+  final Set<SemanticsTag>? tagsForChildren;
```
Can you add a test for the semantics widget to ensure that the tags are properly applied to the tree? (Currently we are only indirectly testing this via the RenderParagraph, but if we ever change how RenderParagraph works, this feature would be untested).

goderbauer reviewed (commented):
LGTM once comments are resolved and @dbrinegar confirmed that this now fixes the crash for the actual use case.

chunhtai committed:
addressing comment

chunhtai commented:
Thanks @goderbauer, I addressed all the comments

@mehmetf @dbrinegar can you verify if this fixes the issue? I have ran the https://github.com/flutter/flutter/issues/69787#issuecomment-722169455 and it fixes the crash

goderbauer reviewed (approved):
LGTM

chunhtai commented:
A gentle ping.

dbrinegar commented on packages/flutter/lib/src/rendering/paragraph.dart:
```diff
@@ -956,6 +988,10 @@ class RenderParagraph extends RenderBox
       }
     }
+    // Makes sure we annotated all of the semantics children.
+    assert(childIndex == children.length);
```
I think I've applied the right patch, but it is now asserting on this line for the case where it wasn't crashing before.

For the case that was crashing, childIndex == children.length == 0, which is where it used to have 7 children expecting 8.  Immediately after that case, childIndex comes out == 0 but children.length == 8 and this asserts.

chunhtai replied:
will reach out offline

dbrinegar replied:
update: patch wasn't right.

Confirmed this fixes the crash.
